src/hb-ot-color-cbdt-table.hh

```diff
@@ -243,6 +435,101 @@ struct IndexSubtableArray
     return_trace (indexSubtablesZ.sanitize (c, count, this));
   }
 
+  bool subset (hb_subset_context_t *c,
+               const char *old_cbdt,
+               hb_vector_t<char> *cbdt,
+               unsigned int *size /* INOUT */,
+               unsigned int *num_tables /* INOUT */,
+               hb_codepoint_t *start_glyph /* OUT */,
+               hb_codepoint_t *end_glyph /* OUT */) const
+  {
+    TRACE_SUBSET (this);
+
+    unsigned int original_num_tables = *num_tables;
+    auto* dst = c->serializer->start_embed<IndexSubtableArray> ();
+    if (unlikely (!dst)) return_trace (false);
+
+    hb_vector_t<hb_pair_t<hb_codepoint_t, const IndexSubtableRecord*>> lookup;
+    for (hb_codepoint_t new_gid = 0; new_gid < c->plan->num_output_glyphs (); new_gid++)
+    {
+      hb_codepoint_t old_gid;
+      if (unlikely (!c->plan->old_gid_for_new_gid (new_gid, &old_gid))) continue;
+
+      const IndexSubtableRecord* record = find_table(old_gid, original_num_tables);
+      if (unlikely (!record)) continue;
+      lookup.push (hb_pair_t<hb_codepoint_t, const IndexSubtableRecord*> (new_gid, record));
+
+      if (*start_glyph > *end_glyph)
+        *start_glyph = new_gid;
+
+      *end_glyph = new_gid;
+    }
+    if (unlikely (!lookup.length)) return_trace (false);
+
+    IndexSubtableRecord *curr_record = nullptr;
+    IndexSubtable *curr_subtable = nullptr;
+    bool success = true;
+    unsigned int num_glyphs = 0;
+    *size = 0;
+    *num_tables = 1;
+    if (unlikely (!lookup[0].second->create_new_record (c->serializer,
+                                                        cbdt->length,
+                                                        this,
+                                                        size,
+                                                        &curr_record,
+                                                        &curr_subtable)))
+      return_trace (false);
+    if (unlikely (!curr_record || !curr_subtable)) return_trace (false);
+
+    for (unsigned int i = 0; i < lookup.length; i++)
+    {
+      // On a match in formats we can extend the old table.
+      const IndexSubtable *old_subtable = lookup[i].second->get_subtable (this);
+      auto* new_header = curr_subtable->get_header ();
+      auto* old_header = old_subtable->get_header ();
+      if (!success ||
+          new_header->indexFormat != old_header->indexFormat ||
+          new_header->imageFormat != old_header->imageFormat)
```

Comment:
Should also check if the current gid (new) is adjacent to the last processed gid (new). If it's not, end the sub table since sub tables should not have gaps in them.

Reply:
Ah good catch, I would do whatever fontTools currently does in this case then. Since we treat that as the gold standard for what the result should be.

Reply:
If the gap exists before subsetting and the glyph for that gap is kept then a space is kept to respect the font's design. fontTools tries to respect the format of existing tables first and foremost so we still need to support gaps.